src/cpu/pred/btb_entry.hh

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2024 Pranith Kumar
+ * All rights reserved.
+ *
+ * The license below extends only to copyright in the software and shall
+ * not be construed as granting a license to any other intellectual
+ * property including but not limited to intellectual property relating
+ * to a hardware implementation of the functionality of the software
+ * licensed hereunder.  You may use the software subject to the license
+ * terms below provided that you ensure that this notice is replicated
+ * unmodified and in its entirety in all distributions of the software,
+ * modified or unmodified, in source code or in binary form.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * Declaration of a BTB entry and BTB indexing policy.
+ */
+
+#ifndef __CPU_PRED_BTB_ENTRY_HH__
+#define __CPU_PRED_BTB_ENTRY_HH__
+
+#include <vector>
+
+#include "arch/generic/pcstate.hh"
+#include "base/cache/cache_entry.hh"
+#include "base/intmath.hh"
+#include "base/types.hh"
+#include "cpu/static_inst.hh"
+#include "mem/cache/tags/indexing_policies/set_associative.hh"
+#include "params/BTBIndexingPolicy.hh"
+
+namespace gem5::branch_prediction
+{
+
+class BTBTagTypes
+{
+  public:
+    struct KeyType
+    {
+        Addr address;
+        ThreadID tid;
+    };
+    using Params = BaseIndexingPolicyParams;
+};
+
+class BTBEntry : public CacheEntry
```

Comment:
I wouldn't inherit from CacheEntry, as this is only adding complexity IMHO with no gain. It forces you to overload and hide some methods. The BTB entry should inherit from ReplaceableEntry

Reply:
I tried inheriting it from ReplaceableEntry and that just seems to be doing most of what CacheEntry is already doing for valid and other methods. We are duplicating most of the code here again.

Is there any reason why you prefer inheriting from ReplaceableEntry instead of CacheEntry and avoid duplication?

Reply:
There are several reasons IMO

1) Design concern: Virtual methods define the interface of a class. If you inherit from a class but you end up overloading such interface, then you are basically saying the BtbEntry is not really a CacheEntry, it is just reusing some of its code.

2) Possible bugs: this hiding can lead to potential bugs, see https://github.com/gem5/gem5/pull/1374

3) Tag ambiguity: the idea of defining a tag as the single address is just an artifact of past design. In this case the tag type is really the pair of addr and tid. But by using the CacheEntry you are basically defining two different tag types.

4) Performance: most of the BTBEntry logic is calling virtual functions. So you are needlessly adding entries to a VTABLE for a polymorphism that is never used, and you are calling them paying the vtable indirection penalty.

I understand that some code will look the same between the CacheEntry and the BTBEntry but IMHO the overhead is minimal and we would benefit a lot in decoupling the two entries, simplifying the design.

Reply:
Thanks for the explanation @giactra. BtbEntry is derived from CacheEntry means it has some properties of CacheEntry, but not all. This is I think a textbook example of OOP (mammals -> humans).

But, I am not really strong against implementing it from scratch. Just that, there will be code duplication.

I pushed a patch addressing all the review comments. Please take a look!